### PR TITLE
Release notes for PS 5.7.23-24

### PR DESCRIPTION
**Pull request by borys-belinsky**

### CODE REVIEW

File: doc/source/management/data_at_rest_encryption.rst [hunk 1]
```diff
@@ -220,18 +192,47 @@ System Variables
...
+
+.. variable:: innodb_redo_log_encrypt
+
+   :version 5.7.23-24: Implemented
+   :cli: ``--innodb-redo-log-encrypt``
+   :dyn: Yes
+   :scope: Global
+   :vartype: Text
+   :default: ``KEYRING_KEY``
+
```
Comment (hrvojem) [inline doc/source/management/data_at_rest_encryption.rst:542]: It seems to be `OFF` by default which isn't documented:
mysql> show variables like 'innodb_redo_log_encrypt';
+-------------------------+-------+
| Variable_name           | Value |
+-------------------------+-------+
| innodb_redo_log_encrypt | off   |
+-------------------------+-------+
1 row in set (0.01 sec)

File: doc/source/management/data_at_rest_encryption.rst [hunk 2]
```diff
@@ -166,12 +138,12 @@ encryption of system temporary tablespace.
   :vartype: Text
   :default: ``OFF``
 
-This feature is considered **BETA** quality.
-
-This variable has 3 possible values. ``ON`` makes |InnoDB| tables encrypted by
+This variable has 5 possible values. ``ON`` makes |InnoDB| tables encrypted by
```
Comment (laurynas-biveinis) [inline doc/source/management/data_at_rest_encryption.rst]: 5 possible, but 3 given in this paragraph

File: doc/source/management/data_at_rest_encryption.rst [hunk 3]
```diff
@@ -220,18 +192,47 @@ System Variables
...
 
+.. _ps.data-at-rest-encryption.redo-log:
+
+Redo Log Encryption
+================================================================================
+
+:Availability: This feature is **Alpha** quality
+
+InnoDB redo log encryption is enabled by setting the variable
+``innodb_redo_log_encrypt``. This variable has two values: ``MASTER_KEY`` and
```
Comment (laurynas-biveinis) [inline doc/source/management/data_at_rest_encryption.rst]: Since the feature is not mandatory, there must be an "off" value as well

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]

File: doc/source/management/data_at_rest_encryption.rst [hunk 4]
```diff
@@ -255,6 +256,23 @@ System Variables
 
 The option turns on encryption of temporary files created by |Percona Server|.
 
+Key Rotation and Scrubbing
```
Comment (dutow) [inline doc/source/management/data_at_rest_encryption.rst]: The title mentions "and scrubbing", but I don't see anyhting about scrubbing here? Also, some part of scrubbing are independent of key rotation.

File: doc/source/release-notes/Percona-Server-5.7.23-24.rst [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+not recommended to be used in production.
+
+New Features
+================================================================================
+
+- :psbug:`4905`: Upstream MyRocks changes have been merged
+- :psbug:`4976`: InnoDB Undo Log Encryption has been implemented
+- :psbug:`4946`: Add the ``rocksdb_no_create_column_family`` option to prevent the implicit creation of column families in MyRocks
+- :psbug:`4556`: InnoDB :ref:`ps.data-at-rest-encryption.redo-log` has been implemented
+- :psbug:`3839`: InnoDB Data Scrubbing has been implemented
```
Comment (dutow) [inline doc/source/release-notes/Percona-Server-5.7.23-24.rst]: PS-3834?

File: doc/source/management/data_at_rest_encryption.rst [hunk 6]
```diff
@@ -220,18 +192,47 @@ System Variables
...
+``MASTER_KEY`` uses the InnoDB master key to encrypt unique keys for each log
+file in the redo log header. ``KEYRING_KEY`` uses the ``percona_redo`` versioned
+key from the keyring. When ``innodb_redo_log_encrypt`` is set to
+``KEYRING_KEY``, each new redo log file is encrypted with the latest
+``percona_redo`` key from the keyring.
+
+System variables
+--------------------------------------------------------------------------------
+
+.. variable:: innodb_redo_log_encrypt
```
Comment (dutow) [inline doc/source/management/data_at_rest_encryption.rst:492]: What about the other variables? Undo log encryption, scrubbing variables for example.

File: doc/source/management/data_at_rest_encryption.rst (hunk 2 above)
Comment (dutow) [reply doc/source/management/data_at_rest_encryption.rst]: Also, I think the new 2 values should be alpha, not beta.

File: doc/source/management/data_at_rest_encryption.rst [hunk 7]
```diff
@@ -255,6 +256,23 @@ System Variables
...
+- CREATE TABLE .. ENCRYPTION='ROTATED_KEYS`
+- ALTER TABLE ... ENCRYPTION='ROTATED_KEYS'
+- CREATE TABLESPACE tablespace_name … ENCRYPTION=’ROTATED_KEYS’
+
+.. note::
+
+   Running ``ALTER TABLE .. ENCRYPTION=’Y’`` on the tablespace created with
+   ``ENCRYPTION=’ROTATED_KEYS’`` converts the table back to the existing MySQL
+   scheme.
+
```
Comment (robgolebiowski) [inline doc/source/management/data_at_rest_encryption.rst:596]: it is not ROTATED_KEYS, but KEYRING

File: doc/source/management/data_at_rest_encryption.rst (hunk 2 above)
Comment (robgolebiowski) [reply doc/source/management/data_at_rest_encryption.rst]: I do not see all the new values ONLINE_TO_KEYRING etc

Comment (robgolebiowski) [reply doc/source/management/data_at_rest_encryption.rst]: In whole docs there is no word about encryption threads

Comment (robgolebiowski) [reply doc/source/management/data_at_rest_encryption.rst]: Or default_encryption_key_id

Comment (robgolebiowski) [reply doc/source/management/data_at_rest_encryption.rst]: This default_encryption_key_id can be altered on tablespace level with ALTER TABLE (…) ENCRYPTION=’KEYRING’ ENCRYPTION_KEY_ID=X.

Comment (robgolebiowski) [reply doc/source/management/data_at_rest_encryption.rst]: what about innodb_encryption_threads and all the other variables ?

**Review by robgolebiowski — CHANGES_REQUESTED** [review]

File: doc/source/management/data_at_rest_encryption.rst [hunk 8]
```diff
@@ -154,34 +273,146 @@ Turning this option off at runtime makes server to create all subsequent
...
+   :version 5.7.21-21: Implemented
+   :cli: ``--innodb-encrypt-tables``
+   :dyn: Yes
+   :scope: Global
+   :vartype: Text
+   :default: ``OFF``
 
-This feature is considered **BETA** quality.
+The implementation of the behavior controlled by this variable is considered
+**ALPHA** quality.
```
Comment (dutow) [inline doc/source/management/data_at_rest_encryption.rst]: The options which are there for several months (ON,OFF,FORCE) should still be beta.

File: doc/source/management/data_at_rest_encryption.rst [hunk 9]
```diff
@@ -154,34 +273,146 @@ Turning this option off at runtime makes server to create all subsequent
...
-default. ``FORCE`` disables creation of unencrypted tables. ``OFF`` restores
-the like-before behavior.
+.. rubric:: FORCE
 
-.. note:: ``innodb_encrypt_tables=ON`` still allows to create unencrypted
-   table with ``ENCRYPTED=NO`` statement, and also allows to create unencrypted
-   general tablespace.
+New tables are created encrypted with the master key. Passing ``ENCRYPTION=NO`` to
+``CREATE TABLE`` or ``ALTER TABLE`` will result in an error and the table will
+not be created or altered.
```
Comment (dutow) [inline doc/source/management/data_at_rest_encryption.rst]: However, altering a table without passing the encryption parameter (and keeping it unencrypted) is possible - so even if it is set to force, it is possible to alter unencrypted tables without encrypting them in the process.

A later part of the doc implies this, but maybe we could mention this here too?

File: doc/source/management/data_at_rest_encryption.rst [hunk 10]
```diff
@@ -154,34 +273,146 @@ Turning this option off at runtime makes server to create all subsequent
...
+      
+.. _data-at-rest-encryption.undo-tablespace:
+
+InnoDB Undo Tablespace Encryption
+================================================================================
+
+.. TODO: More information is needed
+
+The encryption of InnoDB Undo tablespaces is only available when using separate
+undo tablespaces.
```
Comment (dutow) [inline doc/source/management/data_at_rest_encryption.rst]: As without separate undo tablespaces, the undo log is part of the system tablespace, and it's encryption is controlled by the system tablespace encryption option.

And should also be alpha quality?

File: doc/source/management/data_at_rest_encryption.rst [hunk 11]
```diff
@@ -363,33 +683,73 @@ System Variables
...
+
+While data encryption ensures that the existing data are not reachable in
+plain form, the data scrubbing removes the data once the user decides they
+should be deleted. Also, data scrubbing removes copies of data from logs.
+
+Once enabled, data scrubbing works automatically on each tablespace
+separately. To enable data scrubbing, you need to set the following variables:
+
+- :variable:`innodb-background-scrub-data-uncompressed`
+- :variable:`innodb-background-scrub-data-compressed`
```
Comment (dutow) [inline doc/source/management/data_at_rest_encryption.rst:744]: These variables only control tablespace scrubbing, while the paragraph above mentions logs too. Log scrubbing is controlled by a different variable, mentioned earlier. (and similarly to the immediate data scrubbing, it isn't related to the encrpytion threads).

Backround scrubbing also won't do anything when there aren't any encryption threads, as it is executed by those thereads - so it needs that variable >0.

File: doc/source/release-notes/Percona-Server-5.7.23-24.rst (hunk 5 above)
Comment (dutow) [reply doc/source/release-notes/Percona-Server-5.7.23-24.rst]: PS-3829, and any mention of keyring encryption is also missing from here. Unfortunately PS-3829 is set to internal, so it isn't publicly visible. But we should mention keyring encryption here.

File: doc/source/management/data_at_rest_encryption.rst [hunk 12]
```diff
@@ -7,58 +7,207 @@ Data at Rest Encryption
...
+.. code-block:: guess
 
+   mysql> ALTER INSTANCE ROTATE INNODB MASTER KEY
+
+.. rubric:: Doublewrite Buffers
+
+The two types of doublewrite buffers used in |Percona Server| are encrypted
+differently.
+
+When the InnoDB system tablespace is encrypted, you can encrypt the ``doublewrite buffer``
```
Comment (laurynas-biveinis) [inline doc/source/management/data_at_rest_encryption.rst]: not "you can", but that it's just being encrypted as a part of the encrypted system tablespace

File: doc/source/management/data_at_rest_encryption.rst (hunk 10 above)
Comment (laurynas-biveinis) [reply doc/source/management/data_at_rest_encryption.rst]: It is alpha

File: doc/source/management/data_at_rest_encryption.rst [hunk 13]
```diff
@@ -363,33 +683,73 @@ System Variables
...
 connection timeout. Default value is ``15``. Allowed range is from ``1``
 second to ``86400`` seconds (24 hours). The timeout can be also completely
 disabled to wait infinite amount of time by setting this variable to ``0``.
 
-Other reading
--------------
+Data Scrubbing
+================================================================================
+
+While data encryption ensures that the existing data are not reachable in
```
Comment (laurynas-biveinis) [inline doc/source/management/data_at_rest_encryption.rst]: s/reachable/stored ?

In general, the paragraph needs rephrasing. Usual delete for data on disk is marking it as deleted only, and the data itself becomes unreadable only when overwritten. Scrubbing is actively overwriting delete-marked data.

File: doc/source/management/data_at_rest_encryption.rst [hunk 14]
```diff
@@ -363,33 +683,73 @@ System Variables
...
+   :vartype: Boolean
+   :default: ``OFF``
+
+
+.. seealso::
 
-* `Vault Documentation <https://www.vaultproject.io/docs/index.html>`_
-* `General-Purpose Keyring Key-Management Functions
-  <https://dev.mysql.com/doc/refman/5.7/en/keyring-udfs-general-purpose.html>`_
+   Vault Documentation
```
Comment (laurynas-biveinis) [inline doc/source/management/data_at_rest_encryption.rst:779]: Some of the features link to MariaDB docs, some don't, here e.g. https://mariadb.com/kb/en/library/xtradb-innodb-data-scrubbing/

File: doc/source/release-notes/Percona-Server-5.7.23-24.rst (hunk 5 above)
Comment (laurynas-biveinis) [reply doc/source/release-notes/Percona-Server-5.7.23-24.rst]: Let's go with PS-3829 even if not public. We can make it public later

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]
Please also address @dutow comments for the next review round

File: doc/source/management/data_at_rest_encryption.rst (hunk 2 above)
Comment (borys-belinsky) [reply doc/source/management/data_at_rest_encryption.rst]: @robgolebiowski ONLINE_TO_KEYRING and ONLINE_TO_KEYRING_FORCE are these values also *Alpha*?

Comment (laurynas-biveinis) [reply doc/source/management/data_at_rest_encryption.rst]: These two values are also alpha

File: doc/source/management/data_at_rest_encryption.rst (hunk 1 above)
Comment (borys-belinsky) [reply doc/source/management/data_at_rest_encryption.rst:542]: The default has been changed accordingly

File: doc/source/management/data_at_rest_encryption.rst (hunk 2 above)
Comment (borys-belinsky) [reply doc/source/management/data_at_rest_encryption.rst]: The requested values have been added

File: doc/source/management/data_at_rest_encryption.rst (hunk 3 above)
Comment (borys-belinsky) [reply doc/source/management/data_at_rest_encryption.rst]: Default value added

File: doc/source/management/data_at_rest_encryption.rst (hunk 4 above)
Comment (borys-belinsky) [reply doc/source/management/data_at_rest_encryption.rst]: Text restructured with Scrubbing being a separate section

File: doc/source/release-notes/Percona-Server-5.7.23-24.rst (hunk 5 above)
Comment (borys-belinsky) [reply doc/source/release-notes/Percona-Server-5.7.23-24.rst]: Issue added

File: doc/source/management/data_at_rest_encryption.rst (hunk 6 above)
Comment (borys-belinsky) [reply doc/source/management/data_at_rest_encryption.rst:492]: List of variables updated

File: doc/source/management/data_at_rest_encryption.rst (hunk 8 above)
Comment (borys-belinsky) [reply doc/source/management/data_at_rest_encryption.rst]: Status changed as requested

File: doc/source/management/data_at_rest_encryption.rst (hunk 9 above)
Comment (borys-belinsky) [reply doc/source/management/data_at_rest_encryption.rst]: Text updated

File: doc/source/management/data_at_rest_encryption.rst (hunk 10 above)
Comment (borys-belinsky) [reply doc/source/management/data_at_rest_encryption.rst]: Status updated as requested

File: doc/source/management/data_at_rest_encryption.rst (hunk 11 above)
Comment (borys-belinsky) [reply doc/source/management/data_at_rest_encryption.rst:744]: Text updated

File: doc/source/management/data_at_rest_encryption.rst (hunk 12 above)
Comment (borys-belinsky) [reply doc/source/management/data_at_rest_encryption.rst]: Text updated as requested

File: doc/source/management/data_at_rest_encryption.rst (hunk 13 above)
Comment (borys-belinsky) [reply doc/source/management/data_at_rest_encryption.rst]: Text updated

File: doc/source/management/data_at_rest_encryption.rst (hunk 14 above)
Comment (borys-belinsky) [reply doc/source/management/data_at_rest_encryption.rst:779]: @laurynas-biveinis It is not clear how the documentation must be updated in respect to your comment.

File: doc/source/management/data_at_rest_encryption.rst [hunk 15]
```diff
@@ -7,58 +7,207 @@ Data at Rest Encryption
...
 
+The two types of doublewrite buffers used in |Percona Server| are encrypted
+differently.
+
+When the InnoDB system tablespace is encrypted, the ``doublewrite buffer`` pages
+are encrypted as well. The key which was used to encrypt the InnoDB system
+tablespace is also used to encrypt the doublewrite buffer.
+
+|Percona Server| encrypts the ``parallel doublewrite buffer`` with the respective
+tablespace key. Only encrypted tablespace pages are written as encrypted in the
```
Comment (laurynas-biveinis) [inline doc/source/management/data_at_rest_encryption.rst]: "keys" (the parallel doublewrite buffer contain pages from different tablespaces simultaneously, each encrypted with own key)

File: doc/source/management/data_at_rest_encryption.rst [hunk 16]
```diff
@@ -7,58 +7,207 @@ Data at Rest Encryption
...
+Enables the encryption of the parallel doublewrite buffer. For encryption, uses
+the key of the tablespace where the parallel doublewrite buffer is used.
+
+
+.. _innodb_general_tablespace_encryption:
+
+InnoDB General Tablespace Encryption
+================================================================================
+
+.. TODO: add a reference to *existing tablespace support* or reword
```
Comment (laurynas-biveinis) [inline doc/source/management/data_at_rest_encryption.rst]: Please make sure that source code comments are not the sole place where TODOs are tracked - i.e. file JIRA too

File: doc/source/management/data_at_rest_encryption.rst [hunk 17]
```diff
@@ -363,33 +704,81 @@ System Variables
...
--------------
+.. _data-at-rest-encryption.data-scrubbing:
+
+Data Scrubbing
+================================================================================
+
+While data encryption ensures that the existing data are not stored in plain
+form, the data scrubbing literally removes the data once the user decides they
+should be deleted. Compare this behavior with how the ``DELETE`` statement works
+which only marks the affected data as deletable - the space claimed by this data
```
Comment (laurynas-biveinis) [inline doc/source/management/data_at_rest_encryption.rst]: nit: deleted data is not "marked as deletable" (something deleted is already past the point of being deletable), but "marked as deleted"

File: doc/source/management/data_at_rest_encryption.rst (hunk 14 above)
Comment (laurynas-biveinis) [reply doc/source/management/data_at_rest_encryption.rst:779]: How do you think it should be? Should we link to all applicable MariaDB docs or shouldn't we?

File: doc/source/release-notes/Percona-Server-5.7.23-24.rst [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+
+This release introduces InnoDB encryption improvements and merges upstream
+MyRocks changes. Also, the usage of column families in MyRocks has been
+improved. The InnoDB encryption improvements are in **Alpha** quality and are
+not recommended to be used in production.
+
+New Features
+================================================================================
+
+- :psbug:`4905`: Upstream MyRocks changes have been merged
```
Comment (laurynas-biveinis) [inline doc/source/release-notes/Percona-Server-5.7.23-24.rst]: up to "prod201810" tag

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]

File: doc/source/management/data_at_rest_encryption.rst (hunk 14 above)
Comment (borys-belinsky) [reply doc/source/management/data_at_rest_encryption.rst:779]: In text, there is no reference how we use MariaDB implementation. The `see also` block implies that it is "useful reading related to the subject".  I would add more information about MariaDB if we decide to explain the implementation more. in any case it appears to be a separate problem not quite related to the current context of release notes.

File: doc/source/management/data_at_rest_encryption.rst (hunk 16 above)
Comment (borys-belinsky) [reply doc/source/management/data_at_rest_encryption.rst]: The TODO block removed

File: doc/source/management/data_at_rest_encryption.rst (hunk 17 above)
Comment (borys-belinsky) [reply doc/source/management/data_at_rest_encryption.rst]: Text updated as suggested

File: doc/source/release-notes/Percona-Server-5.7.23-24.rst (hunk 18 above)
Comment (borys-belinsky) [reply doc/source/release-notes/Percona-Server-5.7.23-24.rst]: Release notes updated accordingly

File: doc/source/management/data_at_rest_encryption.rst (hunk 15 above)
Comment (borys-belinsky) [reply doc/source/management/data_at_rest_encryption.rst]: Corrected accordingly

**Review by laurynas-biveinis — APPROVED** [review]

**Pull request merged**